Add IRENA stats

energyLS opened the pull request:
## Changes proposed in this Pull Request
This PR adds RE capacity data from IRENA using the powerplantmatching tool, inspired by https://github.com/PyPSA/pypsa-eur/pull/352

Closes #319  

## To do
- [x] Update powerplantmatching in env
- [x] Update mock_snakemake with extra parameter
- [x] Run workflow to get files required for debugging
- [x] Technology mapping does not work with offshore -> replace by zero or nan?
- [x] Debug PyPSA-Eur function
- [x] Remove "TEST_TEMP.py"


## Checklist

- [x] I tested my contribution locally and it seems to work fine.
- [x] Code and workflow changes are sufficiently documented.
- [x] Newly introduced dependencies are added to `envs/environment.yaml` and `envs/environment.docs.yaml`.
- [x] Changes in configuration options are added in all of `config.default.yaml`, `config.tutorial.yaml`, and `test/config.test1.yaml`.
- [x] Changes in configuration options are also documented in `doc/configtables/*.csv` and line references are adjusted in `doc/configuration.rst` and `doc/tutorial.rst`.
- [x] A note for the release notes `doc/release_notes.rst` is amended in the format of previous release notes, including reference to the requested PR.

pz-max commented on config.default.yaml:
```diff
@@ -108,6 +108,22 @@ electricity:
+    # Artificially limit maximum capacities to factor * (IRENA capacities),
+    # i.e. 110% of <years>'s capacities => expansion_limit: 1.1
+    # false: Use estimated renewable potentials determine by the workflow
+    expansion_limit: false
```
@energyLS as I see one application of IRENA stats is to use the stats to define the **expansion** (defines p_nom_max + p_nom_min). This is currently named expansion_limit = true. Since both p_nom_min & max are defined maybe we should rename it to `set_renewable_expansion()`?

The other application of the IRENA stats is to scale the minimum expansion (p_nom_min) of a technology. For instance, to replicate what assets are already installed in today's network as a starting point for the optimization. This happens if I see correctly by default when the config defines `expansion_limit: false`. Now we are missing the option of pure greenfield optimization.

Any idea how we could make these three cases a bit clearer also in the config:
- (set p_nom_min + p_nom_max) * scale factor
- (set p_nom_min) * scale factor
- greenfield (no p_nom_min and no p_nom_max)

energyLS replied:
Hey @pz-max I totally agree, we need to make this clearer esp. in the config.

I see the current options slightly different, or maybe I just put it in another syntax:
- Setting `expansion_limit=false` leads to: `p_nom_min = IRENA_capacities`
- Setting `expansion_limit=scale_factor` additionally leads to: `p_nom_max = p_nom_min * scale factor`
- Greenfield is possible when not calling the function

I do not fully understand what is meant by the expansion `(set p_nom_min + p_nom_max) * scale factor`, it currently does not make sense to me

pz-max replied:
```python  

estimate_renewable_capacities:
    stats: "irena"  # False, = greenfield expansion, 'irena' uses IRENA stats to add expansion limits
    year: 2020  # Available years for IRENA stats 2000 to 2020
    p_nom_min: 1  # any float, scales the minimum expansion from stats
    p_nom_max: false,  # sets the expansion constraint, false, deactivate this option, float scales the p_nom_min factor accordingly 

```

energyLS commented:
@pz-max the PR is ready. Seems like the night coding session was not that bad.
 Some remaining questions from my side:
- The CI has some issues with yamllint, it complains about my yaml files as far as I can understand it. How can I fix that?
- Where exactly can I document changes in made to the config.default.yaml? It shows here "Changes in configuration options are also documented in doc/configtables/*.csv and line references are adjusted in doc/configuration.rst and doc/tutorial.rst." But I might need a short introduction to it
- The feature to compare the IRENA country list with the country list specified by config file is not implemented (yet) because I don't see that this is a common or even rare use case. Nonetheless, I still can implement that

I'd be happy to receive comments on the code, PR in general etc. and show any mistakes or potential improvements!

pz-max commented:
> @pz-max the PR is ready. Seems like the night coding session was not that bad. Some remaining questions from my side:
> 
>  The CI has some issues with yamllint which I can see here (https://results.pre-commit.ci/run/github/343194588/1654782679.rBqJRgrvQw-ROt9Y5r9ccw), it complains about my yaml files as far as I can understand it. How can I fix that?

Can you run pre-commit locally and then try to fix it `pre-commit run --all`. See more detail here:
https://github.com/pypsa-meets-africa/pypsa-africa/blob/a9e9d37cf05c12b146c6ab832a12b144a1cf671d/doc/how_to_contribute.rst#how-to-contribute

>  Where exactly can I document changes in made to the config.default.yaml? It shows here "Changes in configuration options are also documented in doc/configtables/*.csv and line references are adjusted in doc/configuration.rst and doc/tutorial.rst." But I might need a short introduction to it 
 
We didn't document the config options quite well so far. I know how to do it. You need to create a new configtables/<xxx>.csv. Document there all the options and then link it to the doc/configuration.rst. Happy to show you that briefly. A .rst add on will help reading the stuff.

> The feature to compare the IRENA country list with the country list specified by config file is not implemented (yet) because I don't see that this is a common or even rare use case. Nonetheless, I still can implement that

It's a test. Not all config.yaml countries are guaranteed to be correctly named in the IRENA stats (or in future other stats). We faced a ton such naming issues. So applying the check would be quite good.

Let me know when you have some time regarding the config :)

energyLS commented:
I have introduced a new feature to check the country list and fixed the linter issues. I will message you in private for a short discussion on documenting the config

pz-max commented:
Awesome work @energyLS . Is this ready to merge now?

energyLS commented:
> Awesome work @energyLS . Is this ready to merge now?

Thanks! Yes it's ready, although the Linux CI failed when connecting to powerplantmatching. Thats probably just a temporary issue and not in relation with this PR. How can you rerun the Linux CI?

pz-max commented:
> Thanks! Yes it's ready, although the Linux CI failed when connecting to powerplantmatching. Thats probably just a temporary issue and not in relation with this PR. How can you rerun the Linux CI?

Agree. Probably a pp issue. You can rerun the CI by clicking on details and the setting "wheel" -> rerun jobs. Though might be the case that not everyone has permission to do it

energyLS commented:
CI is successfull now! You have probably rerun the jobs. Then the PR is ready to merge

pz-max commented:
Congrats to the great contribution :1st_place_medal:
